src/main/java/com/github/mauricioaniche/ck/CKVisitor.java

```diff
@@ -85,7 +86,7 @@ public boolean visit(TypeDeclaration node) {
 
 		// there might be class level metrics that use the TypeDeclaration
 		// so, let's run them
-		classes.peek().classLevelMetrics.stream().map(metric -> (CKASTVisitor) metric).forEach(ast -> ast.visit(node));
+		classes.peek().classLevelMetrics.stream().map(metric -> (CKASTVisitor) metric).forEach(ast -> {((ClassLevelMetric)ast).setClassName(className); ast.visit(node);});
```

Comment:
Setting the class name here, hmm, that also works! What was the reason to do so?

Reply:
I designed a solution based on the Singleton design pattern for implementing the collection of fan-in and NOC metrics. In the case of fan-in, I have to identify the classes that call a given class. For instance, if I have a class A that calls a class B, I would have to increase 1 for fan-out in A and 1 for fan-in in B. So, I set the class name at this point to identify the class visited at the moment and add it to the fan-in list regarding the classes that it is calling.

In the case of NOC, I do the same thing. But, I observe the call via the keyword **extends** and add the current analyzed class to the NOC list of its superclass.

Reply:
I am now having a better grasp of this `setClassName`. It belongs to the `ClassLevelMetric`. This class represents a metric that is calculated. Setting a class name on it does not really make sense, from a cohesion point of view. The `ClassLevelMetric` is basically an Strategy. Having the name of the class is strange. Also because this means that all strategies will have the name of the current class.

But I understand you need to pass the name of the current class as to then calculate fan in and fan out. 

I'm trying to find a way to avoid this leaking. Do you think it would be better to do it in the `instantiateClassLevelMetricVisitors` method? This method calls the Callable which returns the list of metrics. We'd then manually look for the CBO metric there (look for instanceOf CBO or something) and inject the name of the class only there. It may make this specific method ugly, but it won't hurt the entire metric abstraction...

What do you think?

Reply:
I think trying to pass the class name to the instantiateClassLevelMetricVisitors method instead of the ClassLevelMetric class interface might be a good solution to avoid this leaking. Do you want me to try implementing this strategy here?

Reply:
I can't recall all the details of my previous review, but let's go for it. It will just hurt a single (factory) method instead of the entire class hierarchy!

Reply:
Thank you for your reply, @mauricioaniche! I am revisiting this point, and I will push the changes as soon as possible.

Reply:
Hi, @mauricioaniche! I have already submitted a new commit with the correction that you had asked. Now, the name of the classes are being inserted during instantiation.